source/manual/intro-to-docker-even-more.html.md

```diff
@@ -0,0 +1,398 @@
+```
+
+You should now be able to go to this domain in your browser and see the Content Publisher homepage. After the next section we'll iterate this approach to make it more dynamic and work across multiple apps.
+
+## Step 5: YAML
+
+The `content-publisher-demo` service in our `docker-compose.yml` file is now serving two purposes: running tests and running the app. As we add more environment variables, dependencies, etc., it makes sense to have variants of `content-publisher-demo` that have just the right amount of config for what they need to do.
+
+* Move common config into [an extension][yaml-extension]
+
+```
+# this feature needs a higher version of compose
+version: '3.7'
+
+x-content-publisher-demo: &content-publisher-demo
+  build:
+    context: .
+  volumes:
+    - home:/home/build
+    - ~/govuk:/govuk
+  working_dir: /govuk/content-publisher
+```
+
+* Merge the config hash into the service
+
+```
+services:
+  ...
+
+  content-publisher-demo:
+    <<: *content-publisher-demo
+    ...
+```
+
+* Split the 'demo' service into two 'stacks'
+
+```
+services:
+  ...
+
+  # for tests
+  content-publisher-demo-lite:
+    <<: *content-publisher-demo
+    privileged: true
+    depends_on:
+      - postgres
+    environment:
+      DATABASE_URL: "postgresql://postgres@postgres/content-publisher"
+      TEST_DATABASE_URL: "postgresql://postgres@postgres/content-publisher-test"
+
+  # for Rails
+  content-publisher-demo-app:
+    <<: *content-publisher-demo
+    depends_on:
+      - postgres
+    environment:
+      DATABASE_URL: "postgresql://postgres@postgres/content-publisher"
+      HOST: 0.0.0.0
+    command: bin/rails s
+```
+
+> Although the `depends_on` config is the same for both stacks, in general we expect it to vary between stacks.
+
+* Update the Makefile to use the `lite` stack
+
+Each service will have its own image by default, which is based on the name of the service. Compose will build the images automatically when we try to `run` one of the stacks. Although the build time will be fast, because Docker will re-use layers it built previously, the extra 'build' step is unnecessary and we can easily avoid it.
+
+
+* Build/re-use the same image for both stacks
+
+```
+x-content-publisher-demo: &content-publisher-demo
+  image: content-publisher-demo
+  ...
+```
+
+* Check everything still works after the refactor
+
+```shell
+# re-run setup
+$mac make content-publisher
+
+# run the tests
+$mac docker-compose run content-publisher-demo-lite bundle exec rake
+
+# run the app
+$mac docker-compose run -p 80:3000 content-publisher-demo-app
+```
+
+## Step 6: NGINX
+
+In the previous section we used a `-p 80:3000` flag to make Rails, which runs on port 3000, available on our host machine, on port 80. This approach doesn't scale to multiple apps, since only one process can bind to port 80 at a time. We need a single app that can 'proxy' our HTTP requests...an [`nginx-proxy`][nginx-proxy]!
```

Comment:
```suggestion
In the previous section we used a `-p 80:3000` flag to make Rails, which runs on port 3000, available on our host machine, on port 80. This approach doesn't scale to multiple apps, since only one process can bind to port 80 at a time. We need a single app that can 'proxy' our HTTP requests... an [`nginx-proxy`][nginx-proxy]!
```